src/Datadog.Trace.ClrProfiler.Native/cor_profiler.cpp

```diff
@@ -354,18 +354,31 @@ HRESULT STDMETHODCALLTYPE CorProfiler::AssemblyLoadFinished(AssemblyID assembly_
 
         if (is_instrumentation_assembly)
         {
-            // Configure a version string to compare with the profiler version
+            const auto expected_assembly_reference = AssemblyReference(managed_profiler_full_assembly_version);
+            bool is_viable_version;
+
+            if (runtime_information_.is_core())
+            {
+                is_viable_version = (assembly_metadata.version >= expected_assembly_reference.version);
+            }
+            else
+            {
+                is_viable_version = (assembly_metadata.version == expected_assembly_reference.version);
+            }
+
+            // Configure a version string for logging
             std::stringstream ss;
             ss << assembly_metadata.version.major << '.' << assembly_metadata.version.minor << '.'
                << assembly_metadata.version.build;
 
             auto assembly_version = ToWSTRING(ss.str());
 
-            // Check that Major.Minor.Build match the profiler version
-            if (assembly_version == ToWSTRING(PROFILER_VERSION))
+            // Check that Major.Minor.Build matches the profiler version.
+            // On .NET Core, allow managed library to be a higher version than the native library.
```

Comment:
This is something we don't allow currently right? Do we definitely want to allow that?

While it seems like it would simplify a number of things, it will also significantly reduce the amount of flexibility we have, as it fixes the surface area between managed and native as what we have currently.

e.g. for Telemetry, I'm going to PInvoke into the native lib to get some stats. With this change, I may end up PInvoking into a native lib that doesn't have the method etc. Similarly, the other way, we can't really change any of the managed methods that the native lib calls into, as an "old" native lib could be calling a newer lib (that doesn't have the same signature)

Reply:
Today we require an exact version match between `Datadog.Trace.ClrProfiler.Native.dll` and `Datadog.Trace.ClrProfiler.Managed.dll`, but on Core CLR we could have a newer `Datadog.Trace.dll` if the app is combining an older msi installer with a newer nuget package. This "just works" (if there are no breaking changes in the APIs) because we don't check the version of `Datadog.Trace.dll` anywhere.

With `Datadog.Trace.ClrProfiler.Managed.dll` going away, the native code must now check if `Datadog.Trace.dll` is loaded, but _a newer version of that assembly was already allowed before_, so making the version check more flexible avoids a regression. Example: an app has a newer nuget package and an older msi installed. Auto-instrumentation today loads the newer `Datadog.Trace.dll` and it works. If we leave the strict version check, auto-instrumentation will stop working after this PR because the native dll won't find an assembly with the exact version it is looking for.

Breaking API changes can still happen between versions, we're just moving the boundary where those breaks can happen. 

Before:
```
Datadog.Trace.ClrProfiler.Native.dll ➡️ Datadog.Trace.ClrProfiler.Managed.dll ➡️ (can break here) ➡️ Datadog.Trace.dll
         1.x                     1.x                                 1.y
```
After:
```
Datadog.Trace.ClrProfiler.Native.dll ➡️ (can break here) ➡️ Datadog.Trace.dll
         1.x                                       1.y
```

> it fixes the surface area between managed and native

I think this surface area is smaller and easier to manage the the previous surface area. Also, this tight coupling should go away soon when we stop using `integrations.json` file.

> for Telemetry, I'm going to PInvoke into the native lib to get some stats. With this change, I may end up PInvoking into a native lib that doesn't have the method

That may be something you need to handle anyway, since we can't always guarantee that the version of managed and native libraries will be the same, or that the native library will even be present.

Reply:
I don't think we have a choice but to support it, because in one specific scenario there will only be one version of Datadog.Trace loaded:

- NuGet package: Datadog.Trace v1.n
- Platform package: Datadog.Trace v1.n-1
- Result: All loads return Datadog.Trace v.1.n

We're basically removing possible MissingMethodException's in managed code and replacing it with missing PInvoke calls.